infer/tests/token_test.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"testing"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi-go-provider/integration"
+)
+
+type CustomToken struct{}
+
+func (c *CustomToken) Annotate(a infer.Annotator) { a.SetToken("overwritten", "Tk") }
+
+func (*CustomToken) Create(
+	ctx p.Context, name string, inputs TokenArgs, preview bool,
+) (string, TokenResult, error) {
+	panic("unimplemented")
+}
+
+type TokenArgs struct {
+	Array []ObjectToken `pulumi:"arr"`
+
+	Single ObjectToken `pulumi:"single"`
+}
+type TokenResult struct {
+	Map map[string]ObjectToken `pulumi:"m"`
+}
+
+type TokenComponent struct{ pulumi.ResourceState }
+
+type ComponentToken struct{}
+
+// Check that we allow other capitalization schemes
+func (c *ComponentToken) Annotate(a infer.Annotator) { a.SetToken("cmp", "tK") }
+
+func (*ComponentToken) Construct(
+	ctx *pulumi.Context, name, typ string, inputs TokenArgs, opts pulumi.ResourceOption,
+) (*TokenComponent, error) {
+	panic("unimplemented")
+}
+
+type FnToken struct{}
+
+func (c *FnToken) Annotate(a infer.Annotator) { a.SetToken("fn", "TK") }
+
+func (*FnToken) Call(ctx p.Context, input TokenArgs) (output TokenResult, err error) {
+	panic("unimplemented")
+}
+
+type ObjectToken struct {
+	Value string `pulumi:"value"`
+}
+
+func (c *ObjectToken) Annotate(a infer.Annotator) { a.SetToken("obj", "Customized") }
+
+func TestTokens(t *testing.T) {
```

Comment:
It looks like this change touches the code that generates the default names; do we have a similar test that checks the names produced without these annotations?

Reply:
We have a bunch of integration tests that verify their schema.